### PR TITLE
Fix a problem with multiple APIs clobbering each other in registration.

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/kubernetes/kubernetes/issues/24392

@kubernetes/sig-api-machinery 

[Analytics]()

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28431)

### CODE REVIEW

**Comment by lavalamp** [comment]
Please add a test

**Comment by caesarxuchao** [comment]
@polvi I don't follow your comment in the original issue: "there appears to be a race in kubectl for the ability to read the APIs". I think kubectl is single-threaded, why would there be a race?

I think this PR is correct in that appending filteredGVs to thirdPartyGroupVersions, but I don't understand why @polvi observed on and off error in the original issue, I expect it to always fail.

**Comment by brendandburns** [comment]
@caesarxuchao it's not actually a race.  It's actually because of Golang's randomized map iteration "feature" (I think)

So you load in two APIs on the same path, depending the randomness you iterate them in potentially different orders which makes different APIs found (or not).

I will add a test.

**Comment by caesarxuchao** [comment]
Right, the randomness comes from the map here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L294

**Comment by brendandburns** [comment]
@lavalamp test added, ptal.

Thanks

**Comment by caesarxuchao** [comment]
LGTM. Thanks!

File: pkg/apimachinery/registered/registered.go [hunk 1]
```diff
@@ -70,6 +74,16 @@ func init() {
 	}
 }
 
+// Resets everything to clean room for the start of a test
+func clearForTesting() {
```
Comment (lavalamp) [inline pkg/apimachinery/registered/registered.go]: ugh, this is awful, I really need to get my changes in so it's not necessary :(

**Comment by brendandburns** [comment]
@k8s-bot e2e test this please issue: #27920

**Comment by roberthbailey** [comment]
@brendandburns do you want to cherry pick this into the 1.3 branch?

**Comment by brendandburns** [comment]
I signed it!

**Comment by brendandburns** [comment]
@roberthbailey yeah, I'll prepare a cherry pick once this is in.

**Comment by goltermann** [comment]
@brendandburns @pwittrock This is in v1.4 milestone, but with cherrypick tags.  Is the intent still to get into 1.3.x at some point?

**Comment by brendandburns** [comment]
We're close enough to 1.4 that we can probably wait.

--brendan

________________________________

**Comment by foxish** [comment]
Removing cherrypick tags for now, as this change is already in the 1.4 branch. If the intent is to add it to the 1.3 milestone, please re-add them later.

**Pull request merged**